File: Agent/Globals/Classes/Credits/CreditSpendTerm.py
# One additive term of a CreditSpendRule. Mirrors
# Dock/Globals/Classes/Credits/CreditSpendTerm.js — the cost math MUST stay
# byte-identical across the two services. The term's cost is its `credits`
# coefficient multiplied by the product of (metric / divisor) over every
# dimension it declares. A term with no divisors is flat and always
# contributes its coefficient (this is the ON_START semantics).
#
# `divisors` is keyed by CreditCostDimensions name (e.g. "INPUT_TOKENS").


class CreditSpendTerm:

    def __init__(self, credits: float = 0, divisors: dict = None) -> None:
        self.set_credits(credits)
        self.set_divisors(divisors if divisors is not None else {})

    def get_credits(self) -> float:
        return self.__credits

    def set_credits(self, value) -> None:
        try:
            value = float(value)
        except (TypeError, ValueError):
            value = 0
        self.__credits = value

    def get_divisors(self) -> dict:
        return self.__divisors

    def set_divisors(self, value) -> None:
        sanitized_divisors = {}
        if isinstance(value, dict):
            for dimension_name in value:
                try:
                    divisor = float(value[dimension_name])
                except (TypeError, ValueError):
                    continue
                if divisor > 0:
                    sanitized_divisors[dimension_name] = divisor
        self.__divisors = sanitized_divisors

    def evaluate(self, metrics: dict) -> float:
        amount = self.__credits
        for dimension_name, divisor in self.__divisors.items():
            metric_value = metrics.get(dimension_name, 0) if isinstance(metrics, dict) else 0
            try:
                metric_value = float(metric_value)
            except (TypeError, ValueError):
                metric_value = 0
            amount *= (metric_value / divisor)
        return amount

    def to_json(self) -> dict:
        return {
            "credits": self.get_credits(),
            "divisors": dict(self.get_divisors()),
        }

    @classmethod
    def from_json(cls, data: dict) -> "CreditSpendTerm":
        data = data or {}
        return cls(
            credits=data.get("credits", 0),
            divisors=data.get("divisors", {}),
        )

File: Dock/Globals/Classes/Credits/CreditSpendTerm.js
// One additive term of a CreditSpendRule. The term's cost is its
// `credits` coefficient multiplied by the product of (metric / divisor)
// for every dimension it declares a divisor for. A term with no divisors
// is a flat term — it always contributes its coefficient regardless of
// the metrics, which is what makes ON_START (zero tokens, zero time)
// charge only the flat portion of a rule.
//
// `divisors` is keyed by CreditCostDimensions name (e.g. "INPUT_TOKENS").

class CreditSpendTerm
{
    #credits;
    #divisors;

    constructor({ credits = 0, divisors = {} } = {})
    {
        this.setCredits(credits);
        this.setDivisors(divisors);
    }

    getCredits()
    {
        return this.#credits;
    }

    setCredits(value)
    {
        value = parseFloat(value);
        if (isNaN(value))
        {
            value = 0;
        }
        this.#credits = value;
    }

    getDivisors()
    {
        return this.#divisors;
    }

    setDivisors(value)
    {
        const sanitizedDivisors = {};
        if (value !== null && typeof value === "object")
        {
            for (const dimensionName of Object.keys(value))
            {
                const divisor = parseFloat(value[dimensionName]);
                if (!isNaN(divisor) && divisor > 0)
                {
                    sanitizedDivisors[dimensionName] = divisor;
                }
            }
        }
        this.#divisors = sanitizedDivisors;
    }

    /**
     * Evaluates the term against a metrics object keyed by
     * CreditCostDimensions name. Missing metrics are treated as 0, so a
     * token/time term evaluated at ON_START (empty metrics) yields 0.
     * @param {object} metrics
     * @returns {number}
     */
    evaluate(metrics)
    {
        let amount = this.#credits;

        for (const dimensionName of Object.keys(this.#divisors))
        {
            const divisor = this.#divisors[dimensionName];
            const metricValue = (metrics && typeof metrics[dimensionName] === "number") ? metrics[dimensionName] : 0;
            amount *= (metricValue / divisor);
        }

        return amount;
    }

    toJson()
    {
        return {
            credits: this.getCredits(),
            divisors: { ...this.getDivisors() },
        };
    }

    static fromJson(json)
    {
        return new CreditSpendTerm({
            credits: json?.credits ?? 0,
            divisors: json?.divisors ?? {},
        });
    }
}

module.exports = CreditSpendTerm;
